src/burnchains/affirmation.rs

```diff
@@ -0,0 +1,868 @@
+                if let Some((ref mut confirmed_block_set, ref mut ancestor_burnt)) =
+                    ancestor_confirmations.get_mut(&(*ancestor_height, *ancestor_vtxindex))
+                {
+                    confirmed_block_set.insert(opdata.block_height);
+                    *ancestor_burnt += opdata.burn_fee;
+                } else {
+                    let mut block_set = HashSet::new();
+                    block_set.insert(opdata.block_height);
+                    ancestor_confirmations.insert(
+                        (*ancestor_height, *ancestor_vtxindex),
+                        (block_set, opdata.burn_fee),
+                    );
+                }
+            }
+        }
+    }
+
+    test_debug!("parents = {:?}", &parents);
+    test_debug!("ancestors = {:?}", &ancestors);
+    test_debug!("ancestor_confirmations = {:?}", &ancestor_confirmations);
+
+    if ancestor_confirmations.len() == 0 {
+        // empty prepare phase
+        test_debug!("Prepare-phase has no block-commits");
+        return Ok(None);
+    }
+
+    // find the ancestors with at least $anchor_threshold confirmations, and pick the one that has the
+    // most total BTC.  Break ties by ancestor order -- highest ancestor commit wins.
+    let mut ancestor_block = 0;
+    let mut ancestor_vtxindex = 0;
+    let mut most_burnt = 0;
+    let mut most_confs = 0;
+
+    // consider ancestor candidates in _highest_-first order
+    for ((height, vtxindex), (block_set, burnt)) in ancestor_confirmations.iter().rev() {
+        let confs = block_set.len() as u64;
+        if confs < anchor_threshold.into() {
+            continue;
+        }
+
+        // only consider an earlier ancestor if it burned more than the candidate
+        if *burnt > most_burnt {
+            most_burnt = *burnt;
+            most_confs = confs;
+            ancestor_block = *height;
+            ancestor_vtxindex = *vtxindex;
+        }
+    }
+
+    if most_burnt == 0 {
+        // no anchor block possible -- no block-commit has enough confirmations
+        test_debug!("No block-commit has enough support to be an anchor block");
+        return Ok(None);
+    }
+
+    // find the ancestor that this tip confirms
+    let heaviest_ancestor_header = indexer
+        .read_burnchain_headers(ancestor_block, ancestor_block + 1)?
+        .first()
+        .expect(&format!(
+            "BUG: no block headers for height {}",
+            ancestor_block
+        ))
+        .to_owned();
+
+    let heaviest_ancestor_block =
+        BurnchainDB::get_burnchain_block(burnchain_tx.conn(), &heaviest_ancestor_header.block_hash)
+            .expect(&format!(
+                "BUG: no ancestor block {:?} ({})",
+                &heaviest_ancestor_header.block_hash, heaviest_ancestor_header.block_height
+            ));
+
+    // find the PoX anchor block-commit, if it exists at all
+    // (note that it may not -- a rich attacker can force F*w confirmations with lots of BTC on a
```

Comment:
The block-commit selected here may not correspond to a real Stacks block.  For example, a rich attacker could build a F*w-long chain of block-commits in the prepare-phase and out-spend the rest of the network doing so, only to build off of a block-commit that doesn't even exist.  If such an event occurred, this method would return a false positive -- the block-commit it selects as the anchor block commit would not actually correspond to a Stacks block.  This is okay, though, because ultimately the canonical affirmation map will simply have `a` for this reward cycle.